Added _get_link function and filter only amr detection analysis

mariellemanlulu opened the pull request:
# Description of changes (issue #2 & #11)
## _get_link()
- make use of [irida-uploader's _get_link function](https://github.com/phac-nml/irida-uploader/blob/development/iridauploader/api/api_calls.py#L234) which makes a call to a `target_url` and searches for a `target_key` (a reference link) to return a `href` (the link) of a resource. This allows for easier access to HATEOAS links from IRIDA REST API. 
- Fixes issue #11

## Filter AMR_DETECTION
- Added `get_amr_analysis_results` function:
  1. gets all analysis submission of the given project (raises an exception if the project does not exists)
  2. get all the analysis results of all the analysis submission retrieves
  3. filter only the analysis with `AMR_DETECTION` type (which resides in analysis result object) and return that.
  - This function uses the following helper methods:
    - `_get_analysis_submissions_from_projects`
    - `_get_all_analysis_results`
    - `_is_type_amr`

## Additional changes
- Handled exceptions
- Added exception subclasses

mariellemanlulu committed:
Solve merge conlict when rebasing to development branch.

mariellemanlulu committed:
Added get all analysis results. Renamed function names and variables for clarity.

mariellemanlulu committed:
Clean up for PR. Removed unnecessary functions (get_project_url and get_analysis_submission_url.)

JeffreyThiessen commented on irida_staramr_results/irida_api.py:
```diff
@@ -112,112 +110,169 @@ def _create_session(self):
         """
-        project_analysis_list = []
+        try:
+            project_url = self._get_link(self.base_url + f"/api/projects/{project_id}", "self")
```
A few notes on using `_get_link`
* We should assume that the base_url already includes the `/api/` as part of the base, so don't include it when getting the link, but instead include it when initializing the api
* When adhering to HATEOAS it's better to query in smaller pieces s.t. we fetch projects, then use the target_dict to get the analyses, instead of assuming the url exists where it currently exists. LIke this:
```
project_url = self._get_link(self.base_url, "projects")
analysis_submissions_url = self._get_link(project_url, "project/analyses",
                            target_dict={
                                "key": "identifier",
                                "value": project_id
                            })
```
This should add safety for fetching a project when its possible a user does not have permission to access it.
(I'm not sure the exact code above would work, or if the target dict has analyses links on it, please let me know if/how this works)

mariellemanlulu replied:
The above code works just as you expected. I will change the rest of `_get_link` calls to not assume urls and keep that in mind in the future.

JeffreyThiessen commented on irida_staramr_results/irida_api.py:
```diff
@@ -112,112 +110,169 @@ def _create_session(self):
-                    # add to the list of analysis, by deep copy
-                    project_analysis_list.append(project_analysis.copy())
+            try:
+                analysis_results_url = self._get_link(self.base_url + f"/api/analysisSubmissions/{analysis_submission_id}",
```
Same as above, if possible we should break this into multiple `_get_link` calls using `target_dict`, it will be safely handle the calls.

The additional benefit is because `_get_link` is querying for the identifiers (in this case `analysis_submission_id`) it handles the `KeyError` for us, and generates the `IridaKeyError` for us, raising it to the function caller so we don't even need to have the `try/except` block

mariellemanlulu replied:
Done. I broke it down to multiple _get_link calls.

However, on your next point, I kept the `try/except` block because it is possible for an analysis submission to not have the`"ref": "analysis"` and I needed a way to catch it and handle it by ignoring it. I change the exception type from `KeyError` to `IridaKeyError` to be more specific.

JeffreyThiessen commented on irida_staramr_results/cli.py:
```diff
@@ -47,7 +47,7 @@ def main():
 
     try:
-        analyses = irida.get_analyses_from_projects(args_dict["project"])
+        amr_analyses = irida.get_amr_analysis_results(args_dict["project"])
```
This function can throw a `IridaKeyError` so we need to have an `except IridaKeyError as e:` as part of this `try` to catch it.
With the other changes to `_get_link` usage I am recommending for this PR, it will make `HTTPError`'s not occur, so they could be removed from the `try/except`

mariellemanlulu replied:
Done, I added `except IridaKeyError as e:`. I also found that my error messages were misleading (saying analysis not found instead of project id not found). I fixed that as well.

JeffreyThiessen commented on irida_staramr_results/irida_api.py:
```diff
@@ -112,112 +110,169 @@ def _create_session(self):
-        :param _id:
+        try:
+            project_analysis_submissions = self._get_analysis_submissions_from_projects(project_id)
+        except KeyError:
```
Since we indicated in the function comment that an empty array will be returned if nothing is found, we should return `[]` instead of raising an `IridaKeyError` here.

mariellemanlulu replied:
this IridaKeyError is for catching project Id not existing. I can change the function comment to 

```
Get all AMR detection analysis results from a project id.
If no analysis results found in the project, it returns an empty array.
```

JeffreyThiessen reviewed (changes requested):
There's a few places where we can use `_get_link` to better fit HATEOAS, which will also make the code cleaner/shorter because it will make the errors raised easier to manage.

mariellemanlulu committed:
Code review changes.

Include /api in base_url.

Broke down _get_link calls instead of assuming url.

Fix exception handling.

mariellemanlulu committed:
Merge branch 'development' into get-link-v2

mariellemanlulu committed:
Updated get_amr_analysis_results function comments.

mariellemanlulu committed:
Copy iridauploader api functions for consistency. Issue #11

JeffreyThiessen reviewed (approved):
Looks good, merging
